### ISSUE TITLE
Craftax-Classic: do_action revives despawned and unused mobs on every DO

### ISSUE THREAD

**Issue by JoshuaPurtell** [MichaelTMatthews/Craftax#60]
### Summary

In `craftax/craftax_classic/game_logic.py::do_action`, the player-attack code recomputes every mob mask from health, across all slots:

```python
new_zombies = new_zombies.replace(mask=new_zombies.health > 0)      # line 92
new_cows = new_cows.replace(mask=new_cows.health > 0)               # line 125
new_skeletons = new_skeletons.replace(mask=new_skeletons.health > 0)  # line 158
```

The intent seems to be "the attacked mob dies at health ≤ 0". Because the mask is rebuilt from health for **all** slots, every `DO` action (the whole block is kept when `action == DO`, at line 375) also:

- **Revives despawned mobs.** Despawning clears the mask but keeps health > 0, and inactive mobs keep being moved each step. So despawned zombies and cows reappear wherever they have wandered to, possibly next to the player.
- **Creates zombies from unused slots.** `generate_world` initialises zombie health to 1 (`z_health = jnp.ones(...)`), so the first `DO` of an episode activates all three zombie slots (they start at (0, 0), far from the player, and usually despawn on the next mob update). Cow slots (health `cow_health`) behave the same way.
- **Leaves `mob_map` out of sync.** Revived mobs aren't added to `mob_map`, so until they move they don't block player movement, mob movement, spawning, arrows or block placement.

Skeleton slots start at health 0, so they're only revived after a despawn.

### Reproduction

`craftax==1.6.1` (jax 0.11.2); the same code is on `main` at `c3c2e0d038c4e641f9481320c158f457f30c28f3`. CPU.

```python
import jax, jax.numpy as jnp
from craftax.craftax_classic import game_logic
from craftax.craftax_classic.constants import Action
from craftax.craftax_classic.envs.craftax_symbolic_env import CraftaxClassicSymbolicEnvNoAutoReset

env = CraftaxClassicSymbolicEnvNoAutoReset()
params, static = env.default_params, env.static_env_params
_, state = env.reset_env(jax.random.PRNGKey(0), params)

print(state.zombies.mask, state.cows.mask)          # all False after reset
s = game_logic.do_action(jax.random.PRNGKey(1), state, Action.DO.value, static)
print(s.zombies.mask, s.cows.mask)                  # all True after one DO

# A despawned zombie (mask False, health > 0) near the player comes back on DO:
row, col = int(state.player_position[0]), int(state.player_position[1]) + 5
z = state.zombies.replace(position=state.zombies.position.at[0].set(jnp.array([row, col])),
                          health=state.zombies.health.at[0].set(5),
                          mask=state.zombies.mask.at[0].set(False))
s = game_logic.do_action(jax.random.PRNGKey(2), state.replace(zombies=z), Action.DO.value, static)
print(bool(s.zombies.mask[0]), bool(s.mob_map[row, col]))   # True False
```

Output:

```
[False False False] [False False False]
[ True  True  True] [ True  True  True]
True False
```

### Impact

With a uniform-random policy (2,000 episodes each, 1,200-step horizon), patching this in `do_action` changes episode statistics measurably:

| | median episode length | mean return |
|---|---:|---:|
| as released | 136.5 | 1.46 |
| masks kept for non-target slots | 155 | 1.56 |

A noop policy is unaffected (no `DO`). So agents that use `DO` often, which is almost all of them, face more zombies than the spawn rules alone would produce.

### Suggested fix

Only let a mob's mask go from `True` to `False`, for example:

```python
new_zombies = new_zombies.replace(mask=jnp.logical_and(new_zombies.mask, new_zombies.health > 0))
```

The same applies to cows and skeletons. Clearing `mob_map` for killed mobs (lines ~171–180) already works.

We found this while checking a separate engine implementation against Craftax-Classic with a differential fuzzer. Happy to open a PR if that helps.

**Comment by JoshuaPurtell** [comment]
End-to-end repro of the gameplay impact. It runs uniform-random-policy episodes with the released `do_action`, then with the one-line fix, on the same keys. It only needs `pip install craftax`.

<details><summary>repro_episode_impact.py</summary>

```python
"""End-to-end repro for the Craftax-Classic do_action resurrection bug.

Runs uniform-random-policy episodes with the released `do_action`, then with a
one-line fix (non-target masks may only go True -> False), on the same keys, and
prints episode length and return. Only needs `pip install craftax` (CPU is fine).

    python repro_episode_impact.py [episodes]
"""

import sys

import jax
import jax.numpy as jnp
import numpy as np
from craftax.craftax_classic import game_logic
from craftax.craftax_classic.envs.craftax_symbolic_env import CraftaxClassicSymbolicEnvNoAutoReset

EPISODES = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
HORIZON = 600  # random-policy episodes end well before this
released_do_action = game_logic.do_action


def fixed_do_action(rng, state, action, static_params):
    new = released_do_action(rng, state, action, static_params)
    return new.replace(**{name: getattr(new, name).replace(mask=getattr(new, name).mask & getattr(state, name).mask)
                          for name in ("zombies", "cows", "skeletons")})


def run(label):
    env = CraftaxClassicSymbolicEnvNoAutoReset()
    params = env.default_params

    def episode(key):
        key, reset_key = jax.random.split(key)
        _, state = env.reset_env(reset_key, params)

        def body(carry, _):
            state, key, done, length, ret = carry
            key, act_key, step_key = jax.random.split(key, 3)
            action = jax.random.randint(act_key, (), 0, 17)
            _, nxt, reward, now_done, _ = env.step_env(step_key, state, action, params)
            keep = jnp.logical_not(done)
            state = jax.tree_util.tree_map(lambda a, b: jnp.where(keep, a, b), nxt, state)
            return (state, key, done | now_done, length + keep, ret + jnp.where(keep, reward, 0.0)), None

        (_, _, _, length, ret), _ = jax.lax.scan(body, (state, key, False, 0, 0.0), None, length=HORIZON)
        return length, ret

    keys = jax.random.split(jax.random.PRNGKey(0), EPISODES)
    length, ret = jax.device_get(jax.jit(jax.vmap(episode))(keys))
    print(f"{label:<10} episodes={EPISODES}  median length={np.median(length):6.1f}  "
          f"mean length={length.mean():6.1f}  mean return={ret.mean():.3f}")


run("released")
game_logic.do_action = fixed_do_action  # craftax_step looks it up from the module at trace time
run("fixed")
```
</details>

Output (`python repro_episode_impact.py 1000`, craftax 1.6.1, jax 0.11.2, CPU):

```
released   episodes=1000  median length= 140.0  mean length= 146.2  mean return=1.440
fixed      episodes=1000  median length= 159.0  mean length= 158.6  mean return=1.542
```

The minimal repro from the issue, re-run on the same setup:

```
after reset                  zombies.mask=[False, False, False] cows.mask=[False, False, False] zombies.health=[1, 1, 1]
after do_action(DO)          zombies.mask=[True, True, True] cows.mask=[True, True, True] zombies.health=[1, 1, 1]
after do_action(NOOP)        zombies.mask=[False, False, False] cows.mask=[False, False, False] zombies.health=[1, 1, 1]

despawned zombie at (32, 37): mask before DO = False, after DO = True, mob_map at its tile = False
```

(The first version of this issue gave the craftax version as 0.11.2; that was the jax version. It's corrected above.)

**Comment by JoshuaPurtell** [comment]
^^ fully AI generated. but this comment is not!
